File: TopCoder/src/main/java/com/leetcode/String/ShortestDistanceToAChar.java
package com.leetcode.String;

//https://leetcode.com/problems/shortest-distance-to-a-character/discuss/125788/C++JavaPython-2-Pass-with-Explanation
public class ShortestDistanceToAChar {

    public int[] shortestToChar(String S, char C) {
        int n = S.length();
        int[] res = new int[n];
        int pos = -n;

        for (int i = 0 ; i < S.length() ; i++) {
            if (S.charAt(i) == C) {
                pos = i;
            }
            res[i] = i - pos;
        }

        for ( int i = S.length() - 1 ; i >= 0 ; i--) {
            if (S.charAt(i) == C) {
                pos = i;
            }
            res[i] = Math.min(res[i], Math.abs(pos - i));
        }
        return res;
    }

    public static void main(String[] args) {
        String S = "loveleetcode";
        char C = 'e';
        int[] res = new ShortestDistanceToAChar().shortestToChar(S, C);
        for ( int i = 0 ; i < res.length ; i++) {
            System.out.println(res[i]);
        }
    }

}

File: TopCoder/python/misc/ShortestDistanceToAChar.py

class Solution:
    """
    def shortestToChar(self, S, C):
        res = [1 for x in range(len(S))]
        n = len(S)
        pos = -n
        for i in range(len(S)):
            if S[i] == C:
                pos = i

            res[i] = i - pos

        for i in range(len(S)-1, -1, -1):
            if S[i] == C:
                pos = i

            res[i] = min(res[i], abs(pos - i))

        return res
    """

    def shortestToChar(self, S, C):
        result = []

        indexes = []

        for i in range(len(S)):
            if S[i] == C:
                indexes.append(i)
        index = 0
        lower = indexes[index]
        upper = indexes[index]

        for i in range(len(S)):
            if i < lower:
                result.append(lower - i)
            elif lower < i < upper:
                result.append(min(i-lower, upper-i))
            elif lower > upper:
                result.append(i-upper)
            elif i == lower:
                if index + 1 < len(indexes):
                    lower = upper
                    result.append(0)
                    index = index+1
                    upper = indexes[index]
                else:
                    upper = lower
            elif i == upper:
                result.append(0)
                if index + 1 < len(indexes):
                    lower = upper
                    index = index+1
                    upper = indexes[index]
                else:
                    lower = upper
        return result

S = "loveleetcode"
C = 'e'

sol = Solution()
resultNew = sol.shortestToChar(S, C)

print(resultNew)
